keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+
+There is going to be a controller to avoid leaking resources:
+
+- checking that the corresponding parentReference on the IPAddress match the corresponding Service
+- cleaning all IPAddress without an owner reference or, if the time since it was created is greater
+than 60 seconds (default timeout value on the kube-apiserver )
+
+#### Edge cases
+
+Since we have to maintain 3 relationships Services, ServiceCIDRConfigs and IPAddresses, we should be able
+to handle edge cases and race conditions.
+
+- Valid Service and IPAddress without ServiceCIDRConfig:
+
+This situation can happen if a user forcefully delete the ServiceCIDRConfig, it can be recreated for the
+"default" ServiceCIDRConfigs because the information is in the apiserver flags, but for other ServiceCIDRConfigs
+that information is no longer available.
+
+Another possible situation is when one ServiceCIDRConfig has been deleted, but the information takes too long
+to reach one of the apiservers, its allocator will still consider the range valid and may allocate one IP
+from there. To mitigate this problem, we'll set a grace period of 60 seconds on the servicecidrconfig controller
+to remove the finalizer, if an IP address is created during this time we'll be able to block the deletion
+and inform the user.
+
+[![](https://mermaid.ink/img/pako:eNp1kjFvwyAQhf8KYm3ixlZVtQyRbNwhW9WsLCe4JEg2uBi3qqL890Jw67hNGBDifffuieNIpVVIGe3xfUAjsdawd9AKQ8IC6a0jFYGe1NigR7JF96El8k39xq3Z6X0CO3BeS92B8YRHHDrdBxQdKf8T9ZyoLpuVUeMOYWomTAIqslyv7zi7mYXkz0WWPz5lq2x1XzykqrSXsZbU7I_1qBobrmzMEghoGisjM7nlReLqs0vJfsuTm0oqj-qyYleCqXPiiRvDzPPOqSnTRb_NK_nU_mAHf3USwtAFbdG1oFWY6TE6CeoP2KKgLBwV7mBovKDCnAI6dCrEf1E6vDxlO2h6XFAYvN1-GUmZdwP-QOO_GKnTN31Vsn4)](https://mermaid.live/edit#pako:eNp1kjFvwyAQhf8KYm3ixlZVtQyRbNwhW9WsLCe4JEg2uBi3qqL890Jw67hNGBDifffuieNIpVVIGe3xfUAjsdawd9AKQ8IC6a0jFYGe1NigR7JF96El8k39xq3Z6X0CO3BeS92B8YRHHDrdBxQdKf8T9ZyoLpuVUeMOYWomTAIqslyv7zi7mYXkz0WWPz5lq2x1XzykqrSXsZbU7I_1qBobrmzMEghoGisjM7nlReLqs0vJfsuTm0oqj-qyYleCqXPiiRvDzPPOqSnTRb_NK_nU_mAHf3USwtAFbdG1oFWY6TE6CeoP2KKgLBwV7mBovKDCnAI6dCrEf1E6vDxlO2h6XFAYvN1-GUmZdwP-QOO_GKnTN31Vsn4)
+
+For any Service and IPAddress that doesn't belong to a ServiceCIDRConfig the controller will raise an event
+informing the user, keeping the previous behavior
+
+```go
+// cluster IP is out of range
+c.recorder.Eventf(&svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]:%s is not within the service CIDR %s; please recreate service",
+```
+
+- Valid Service and ServiceCIDRConfig but not IPAddress
+
+It can happen that an user forcefully delete an IPAddress, in this case, the controller will regenerate the IPAddress.
+
+During this time, there is a chance that an apiserver tries to allocate this IPAddress, with a possible situation where
+2 Services has the same IPAddress. In order to avoid it, the Allocator will not delete an IP from its local cache
+until it verifies that the consumer associated to that IP has been deleted too.
```

Comment:
I forget if apiserver ones run repair loop at startup?  If it does, that's enough to rebuild the cache and (at least for services) prevent duplicate allocation.

I think

Reply:
they do, there is a bootstrap controller that run in a post-start hook, this controller run the repair loops.